### PR TITLE
change: adding error types

### PR DESCRIPTION
**Pull request by krneta**
*Issue #, if available:*

*Description of changes:*

Related to : https://github.com/amazon-braket/BraketSimulator.jl/pull/27

*Testing done:*

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [ ] I have read the [CONTRIBUTING](https://github.com/amazon-braket/amazon-braket-simulator-v2-python/blob/main/CONTRIBUTING.md) doc
- [ ] I used the commit message format described in [CONTRIBUTING](https://github.com/amazon-braket/amazon-braket-simulator-v2-python/blob/main/CONTRIBUTING.md#commit-your-change)
- [ ] I have updated any necessary documentation, including [READMEs](https://github.com/amazon-braket/amazon-braket-simulator-v2-python/blob/main/README.md) and [API docs](https://github.com/amazon-braket/amazon-braket-simulator-v2-python/blob/main/CONTRIBUTING.md#documentation-guidelines) (if appropriate)

#### Tests

- [ ] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [ ] I have checked that my tests are not configured for a specific region or account (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/braket/simulator_v2/base_simulator_v2.py [hunk 1]
```diff
@@ -11,6 +13,7 @@
 from braket.ir.openqasm import Program as OpenQASMProgram
 from braket.simulator import MultiSimulator
 from braket.task_result import GateModelTaskResult
+from juliacall import JuliaError
```
Comment (kshyatt-aws) [inline src/braket/simulator_v2/base_simulator_v2.py:16]: Can we combine this with line 3?

File: test/unit_tests/braket/simulator_v2/test_density_matrix_simulator_v2.py [hunk 2]
```diff
@@ -676,7 +675,7 @@ def test_simulator_fails_samples_0_shots():
     )
     with pytest.raises(ValueError):
         simulator.run(jaqcd, qubit_count=1, shots=0)
-    with pytest.raises(juliacall.JuliaError):
+    with pytest.raises(ValueError):
```
Comment (kshyatt-aws) [inline test/unit_tests/braket/simulator_v2/test_density_matrix_simulator_v2.py:678]: Can this and the test right above be merged?

**Review by kshyatt-aws — APPROVED** [review]

File: src/braket/simulator_v2/base_simulator_v2.py [hunk 3]
```diff
@@ -103,7 +100,10 @@ def run_jaqcd(
                 qubit_count,
             )
 
-        r = jl.simulate(self._device, self._jaqcd_to_jl(circuit_ir), qubit_count, shots)
+        try:
+            r = jl.simulate(self._device, self._jaqcd_to_jl(circuit_ir), qubit_count, shots)
+        except juliacall.JuliaError as e:
```
[the remark this answers is not in the document]
Comment (krneta) [reply src/braket/simulator_v2/base_simulator_v2.py:105]: Yes, it won't get caught if it's a different type of exception (it'll keep bubbling up).

File: test/unit_tests/braket/simulator_v2/test_density_matrix_simulator_v2.py (hunk 2 above)
Comment (krneta) [reply test/unit_tests/braket/simulator_v2/test_density_matrix_simulator_v2.py:678]: It's generally better/cleaner to have failure-case tests separate from happy-case tests. Besides, it's probably also a good idea to try to keep these files as similar as possible to the default sim, to make them easier to sync (if we need to).

File: test/unit_tests/braket/simulator_v2/test_state_vector_simulator_v2.py [hunk 4]
```diff
@@ -816,7 +808,7 @@ def test_simulator_run_amplitude_no_shots_invalid_states():
     )
     with pytest.raises(ValueError):
         simulator.run(jaqcd, qubit_count=2, shots=0)
-    with pytest.raises(juliacall.JuliaError):
+    with pytest.raises(ValueError):
```
[the remark this answers is not in the document]
Comment (krneta) [reply test/unit_tests/braket/simulator_v2/test_state_vector_simulator_v2.py:811]: Do you mean the "with" lines? Each one raises an error, so they need to be caught separately (or only one would actually run).

[the remark this answers is not in the document]
Comment (kshyatt-aws) [reply test/unit_tests/braket/simulator_v2/test_state_vector_simulator_v2.py:811]: I guess what I mean is, could this be moved to a top level pytest parametrize? Cool if not, was just curious

**Pull request merged**